### PR TITLE
Rework concurrency in the profile manager.

### PR DESCRIPTION
**Pull request by charlesmchen**
PTAL @michaelkirk

### CODE REVIEW

**Comment by charlesmchen** [comment]
Ignoring whitespace will clarify: https://github.com/WhisperSystems/Signal-iOS/pull/2433/files?w=1

**Comment by charlesmchen** [comment]
There's a lot more `synchronized` now which could block some reads by the UI but profile-related writes should be rare and narrow, so I'm not too worried about it.

File: Signal/src/Profiles/OWSProfileManager.m [hunk 1]
```diff
@@ -101,26 +95,26 @@ @interface OWSProfileManager ()
 @property (nonatomic, readonly) YapDatabaseConnection *dbConnection;
 @property (nonatomic, readonly) TSNetworkManager *networkManager;
 
+// These properties can be accessed on any thread, while synchronized on self.
```
Comment (michaelkirk) [inline Signal/src/Profiles/OWSProfileManager.m]: nit: Previously while reading comments making statements about "these properties" I've been confused as to what was actually included in "these properties".  Consider either making a line comment per property or use the specific property names in the comment.

File: Signal/src/Profiles/OWSProfileManager.m [hunk 2]
```diff
@@ -225,150 +219,170 @@ - (UserProfile *)getOrBuildUserProfileForRecipientId:(NSString *)recipientId
...
+        BOOL isLocalUserProfile = userProfile == self.localUserProfile;
+
+        dispatch_async(dispatch_get_main_queue(), ^{
+            if (isLocalUserProfile) {
+                [[NSNotificationCenter defaultCenter] postNotificationName:kNSNotificationName_LocalProfileDidChange
+                                                                    object:nil
+                                                                  userInfo:nil];
+            } else {
+                [[NSNotificationCenter defaultCenter]
+                    postNotificationName:kNSNotificationName_OtherUsersProfileDidChange
```
Comment (michaelkirk) [inline Signal/src/Profiles/OWSProfileManager.m]: seems likely we'll want to include the recipientId or the userProfile in the notification.

File: Signal/src/Profiles/OWSProfileManager.m [hunk 3]
```diff
@@ -225,150 +219,170 @@ - (UserProfile *)getOrBuildUserProfileForRecipientId:(NSString *)recipientId
...
-        NSString *_Nullable avatarUrlPath, NSString *_Nullable avatarFileName) {
-        [self updateServiceWithProfileName:profileName
-            success:^{
-                // All reads and writes to user profiles should happen on the main thread.
+    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+        @synchronized(self)
+        {
+            // Ensure that the success and failure blocks are called on the main thread.
+            void (^failureBlock)() = ^{
                 dispatch_async(dispatch_get_main_queue(), ^{
```
Comment (michaelkirk) [inline Signal/src/Profiles/OWSProfileManager.m]: For my own edification, do you have a preference for using this style vs the `DispatchMainThreadSafe` ?

File: Signal/src/Profiles/OWSProfileManager.m [hunk 4]
```diff
@@ -671,156 +704,181 @@ - (void)setContactRecipientIds:(NSArray<NSString *> *)contactRecipientIds
...
+                            } else if (error) {
+                                DDLogError(@"%@ avatar download failed: %@", self.tag, error);
+                            } else if (!encryptedData) {
+                                DDLogError(@"%@ avatar encrypted data could not be read.", self.tag);
+                            } else if (!decryptedData) {
+                                DDLogError(@"%@ avatar data could not be decrypted.", self.tag);
+                            } else if (!image) {
+                                DDLogError(@"%@ avatar image could not be loaded: %@", self.tag, error);
+                            } else {
+                                // TODO: Verify that the avatar URL hasn't changed since the download began.
```
Comment (michaelkirk) [inline Signal/src/Profiles/OWSProfileManager.m]: re: TODO, it seems like you've addressed this with your additions a few lines up?

```
 } else if (![avatarUrlPathAtStart isEqualToString:currentUserProfile.avatarUrlPath]) {
...
```

Also to nit: `currentUserProfile` isn't the best name. For a second I thought it referred to the "local" user. Maybe `latestUserProfile`? Names are hard. Take it or leave it.

File: Signal/src/Profiles/OWSProfileManager.m (hunk 1 above)
Comment (charlesmchen) [inline Signal/src/Profiles/OWSProfileManager.m]: Yup. I've tried to make things clearer by adopting a single policy: "only access the mutable state while synchronized on self".  Anyway, I've added per-property comments, but they're all the same now.

File: Signal/src/Profiles/OWSProfileManager.m (hunk 2 above)
Comment (charlesmchen) [inline Signal/src/Profiles/OWSProfileManager.m]: Yup!  For whoever does the "update views to reflect changes" work.

File: Signal/src/Profiles/OWSProfileManager.m (hunk 3 above)
Comment (charlesmchen) [inline Signal/src/Profiles/OWSProfileManager.m]: DispatchMainThreadSafe() is an optimization - it avoids dispatching async if we're already on the main thread.  It comes at the cost of being slightly harder to read/understand.  I try to only use it where we need that optimization (e.g. to update a view synchronously if possible).

File: Signal/src/Profiles/OWSProfileManager.m (hunk 4 above)
Comment (charlesmchen) [inline Signal/src/Profiles/OWSProfileManager.m]: Done and done.

**Pull request merged**